File: Solutions/74_Search_a_2D_Matrix/74_Search_a_2D_Matrix.py
class Solution:
    def searchMatrix(self, matrix, target):
        if not matrix or not matrix[0]: return False
        m, n = len(matrix[0]), len(matrix)
        beg, end = 0, m*n - 1
        while beg < end:
            mid = (beg + end)//2
            if matrix[mid//m][mid%m] < target:
                beg = mid + 1
            else:
                end = mid
        return matrix[beg//m][beg%m] == target

#Third fastest
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if not matrix:
            return 0
        
        m=len(matrix)
        n=len(matrix[0])
        l=0
        r=m*n-1
        
        
        while l<=r:
            mid=(l+r)//2
            nums=matrix[mid//n][mid%n]
            if nums==target:
                return True
            elif nums<target:
                l=mid+1
            else:
                r=mid-1
        return False

#Second fastest
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if not matrix:
            return 0
        elif not matrix[0]:
            return 0
        firstCol = []
        for row in matrix:
            firstCol.append(row[0])
        rowId = bisect.bisect(firstCol, target) - 1
        if matrix[rowId][0] == target:
            return True
        colId = bisect.bisect(matrix[rowId], target) - 1
        #print(rowId, colId)
        try:
            if matrix[rowId][colId] == target:
                return True
            else:
                return False
        except:
            return False


#Fastest
class Solution:
    def searchMatrix(self, ma: List[List[int]], target: int) -> bool:
        if len(ma)>0:
            n=len(ma)
            m=len(ma[0])

            low=0
            high=n*m-1
            while low<=high:
                mid=low+(high-low)//2
                if ma[mid//m][mid%m]==target:
                    return True
                elif ma[mid//m][mid%m]>target:
                    high=mid-1
                elif ma[mid//m][mid%m]<target:
                    low=mid+1
            else:
                return False
        else:
            return False

File: Solutions/74_Search_a_2D_Matrix/74_Search_a_2D_Matrix.go
func searchMatrix(matrix [][]int, target int) bool {
    if len(matrix) < 1 {
        return false
    }
    n := len(matrix[0])
    if n < 1{
        return false
    }
    low, hi := 0, len(matrix) - 1
    for low <= hi {
        m := (low+hi)/2
        if matrix[m][0] <= target && target <= matrix[m][n-1]{
            colLow, colHi := 0, n-1
            for colLow <= colHi {
                colM := (colLow + colHi)/2
                if target == matrix[m][colM]{
                    return true
                } 
                if target < matrix[m][colM]{
                    colHi = colM -1
                } else{
                    colLow = colM +1
                }
                
            }
        }
        if matrix[m][0] > target{
            hi = m -1
        } else{
            low = m + 1
        }
    }
    
    return false
}

//Another
func searchMatrix(matrix [][]int, target int) bool {
	columnMatrix := len(matrix)
	if columnMatrix == 0 {
		return false
	}
	rowMartix := len(matrix[0])
	if rowMartix == 0 {
		return false
	}

	xMatrix := 0
	yMatrix := columnMatrix - 1

	for yMatrix >= 0 && xMatrix < rowMartix {
		if matrix[yMatrix][xMatrix] == target {
			return true
		} else if matrix[yMatrix][xMatrix] < target {
			xMatrix++
		} else if matrix[yMatrix][xMatrix] > target {
			yMatrix--
		}
	}
	return false
}